### PR TITLE
arm/coproc: Start coproc development

### PR DESCRIPTION
**Pull request by otyshchenko1**
It is my first commit. Please, don't rely on it too much since I'll likely change it soon. 

Signed-off-by: Oleksandr Tyshchenko olekstysh@gmail.com

### CODE REVIEW

**Comment by otyshchenko1** [comment]
ok

File: xen/arch/arm/coproc/coproc.h [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    char *name;
+    struct device *dev;
+
+    u32 num_mmios;
+    struct mmio *mmios;
+    u32 num_irqs;
+    unsigned int *irqs;
+    struct list_head list;
+
+    const struct vcoproc_instance_ops *ops;
```
Comment (aanisov) [inline xen/arch/arm/coproc/coproc.h]: please rename the structure to vcoproc_ops

File: xen/arch/arm/coproc/coproc.h [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+#include <xen/sched.h>
+
+struct mmio {
+    void __iomem *base;
+    u64 addr;
+    u64 size;
+};
+
+struct coproc_device {
```
Comment (aanisov) [inline xen/arch/arm/coproc/coproc.h]: Please add a list of vcoprocs associated to this device

File: xen/arch/arm/coproc/coproc.c [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+#include <xen/list.h>
+#include <asm/device.h>
+
+#include "coproc.h"
+
+static DEFINE_SPINLOCK(coproc_devices_lock);
+static LIST_HEAD(coproc_devices);
+static int num_coprocs_devices;
+
+int vcoproc_register(struct domain *d, struct vcoproc_info *info)
```
Comment (aanisov) [inline xen/arch/arm/coproc/coproc.c]: IMHO, a function with such f name should associate a vcoproc to a coproc device, in other word register a vcoproc within the coproc framework. NOT link the vcoproc to a domain, which is pure vcoproc_domian function

File: xen/arch/arm/coproc/coproc.h [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+ * GNU General Public License for more details.
+ */
+
+#ifndef __COPROC_H_
+#define __COPROC_H_
+
+#include <xen/sched.h>
+
+struct mmio {
+    void __iomem *base;
```
Comment (aanisov) [inline xen/arch/arm/coproc/coproc.h]: Why do we need this pointer here?

Comment (otyshchenko1) [inline xen/arch/arm/coproc/coproc.h]: The "base" is what ioremap_nocache() returns. It is needed for accessing coproc IO registers.

File: xen/arch/arm/coproc/coproc.h (hunk 1 above)
Comment (otyshchenko1) [inline xen/arch/arm/coproc/coproc.h]: ок

File: xen/arch/arm/coproc/coproc.h (hunk 2 above)
Comment (otyshchenko1) [inline xen/arch/arm/coproc/coproc.h]: ok

File: xen/arch/arm/coproc/coproc.h [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+#ifndef __COPROC_H_
+#define __COPROC_H_
+
+#include <xen/sched.h>
```
Comment (otyshchenko1) [inline xen/arch/arm/coproc/coproc.h]: I am afraid but I don't understand you point here.
sched.h is located in include/xen folder.

File: xen/arch/arm/coproc/coproc.c (hunk 3 above)
Comment (otyshchenko1) [inline xen/arch/arm/coproc/coproc.c]: I'll think about it

**Comment by otyshchenko1** [comment]
OK. The main part of your comments I am going to address soon.

File: xen/arch/arm/coproc/plat/coproc_xxx.c [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+
+	spin_lock(&coproc_xxx->vcoprocs_lock);
+	list_add(&vcoproc_xxx->list, &coproc_xxx->vcoprocs_list);
+	spin_unlock(&coproc_xxx->vcoprocs_lock);
+
+	ret = vcoproc_attach(d, vcoproc_xxx);
+	if (ret) {
+		dev_err(coproc_xxx->dev, "failed to attach vcoproc to domain %d\n", d->domain_id);
+		return ret;
+	}
```
Comment (aanisov) [inline xen/arch/arm/coproc/plat/coproc_xxx.c]: first attach the vcoproc to the domain then register vcoproc in coproc framework
or should take care to cleanup registration on failed attachement

File: xen/arch/arm/coproc/coproc.c [hunk 7]
```diff
@@ -23,6 +23,38 @@ static DEFINE_SPINLOCK(coproc_devices_lock);
...
 
+int vcoproc_context_switch(struct vcoproc_info *prev, struct vcoproc_info *next)
+{
+    struct coproc_device *coproc;
+    int ret = 0;
+
+    if ( unlikely(prev == next) )
+        return 0;
+
+    coproc = next ? next->coproc : prev->coproc;
```
Comment (aanisov) [inline xen/arch/arm/coproc/coproc.c]: What is the purpose of this line?

File: xen/arch/arm/coproc/coproc.h [hunk 8]
```diff
@@ -47,25 +49,86 @@ struct vcoproc_info {
     /* list is used to append instances of vcoproc to vcoprocs_list */
     struct list_head list;
 
+    /* TODO */
+    bool_t is_running;
```
Comment (aanisov) [inline xen/arch/arm/coproc/coproc.h]: I would introduce a state with macro at the moment, 'cause we likely would have number of states

File: xen/arch/arm/coproc/coproc.c (hunk 7 above)
Comment (otyshchenko1) [inline xen/arch/arm/coproc/coproc.c]: to fill a pointer with a valid value

File: xen/arch/arm/coproc/coproc.h (hunk 8 above)
Comment (otyshchenko1) [inline xen/arch/arm/coproc/coproc.h]: ok. I will introduce it later.

File: xen/arch/arm/coproc/coproc.h [hunk 9]
```diff
@@ -47,25 +49,86 @@ struct vcoproc_info {
...
 };
 
 struct vcoproc_domain_ops {
     int (*domain_init)(struct domain *, struct vcoproc_info *);
     void (*domain_free)(struct domain *, struct vcoproc_info *);
 };
 
+/* TODO - move all scheduler stuff out of this header */
+
+struct vcoproc_task_slice {
```
Comment (aanisov) [inline xen/arch/arm/coproc/coproc.h]: what is this?

**Pull request closed without merging**